source/common/stats/thread_local_store.cc

```diff
@@ -75,6 +91,36 @@ void ThreadLocalStoreImpl::shutdownThreading() {
   shutting_down_ = true;
 }
 
+void ThreadLocalStoreImpl::mergeHistograms() {
+  if (!shutting_down_) {
+    std::unique_lock<std::mutex> lock(lock_);
+    // TODO: Wait till all threads have been checked-in
+    tls_->runOnAllThreads([this]() -> void {
+      for (ScopeImpl* scope : scopes_) {
+        for (auto histogram : tls_->getTyped<TlsCache>().scope_cache_[scope].histograms_) {
+          TlsHistogramSharedPtr tls_histogram_ptr =
+              std::dynamic_pointer_cast<ThreadLocalHistogramImpl>(histogram.second);
+          tls_histogram_ptr->beginMerge();
+        }
+      }
+    });
+    if (main_thread_dispatcher_) {
+      main_thread_dispatcher_->post([this]() -> void { mergeInternal(); });
+    }
+  }
+}
+
+void ThreadLocalStoreImpl::mergeInternal() {
+  std::unique_lock<std::mutex> lock(lock_);
+  for (ScopeImpl* scope : scopes_) {
+    for (auto histogram : scope->central_cache_.histograms_) {
+      ParentHistogramSharedPtr parent_hist_ptr =
+          std::dynamic_pointer_cast<HistogramParentImpl>(histogram.second);
```

Comment:
to remove all the dynamic casts here, I would change the map to store impls so you can use them locally.